tenantsettingswatcher: implement watcher and integrate into server

RaduBerinde opened the pull request:
~For now, we hardcode 50 as the system table ID. I plan to work on that and add a commit to this PR. Adding do-not-merge label until then.~

#### tenantsettingswatcher: in-memory data structure

This commit introduces the overrides store, the in-memory data
structure that will hold all tenant overrides.

The data structure is designed for efficient read access and to allow
multiple goroutines to get notified of changes (via a channel close).

Release note: None

#### tenantsettingswatcher: implement watcher

This commit implements the tenantsettingswatcher, which monitors the
tenant_settings table and allows callers to access the current
overrides and listen for changes.

For now, we hardcode the tenant_settings system table ID to 50 (which
will be the case in fresh clusters). We will fix this in a follow-up
change.

Release note: None

#### tenantsettingswatcher: integrate into server

This commit integrates the tenant settings watcher into the server and
implements the TenantSettings API.

Release note: None

#### logictestccl: add tenant_settings test

This commit adds functionality to logic tests to allow running SQL
against the host cluster in a tenant configuration. This allows us to
manipulate the tenant_settings table and observe the settings changing
on the tenant.

We add a tenant_settings test that verifies basic overrides behavior.
This test can become more comprehensive once we implement the new
ALTER TENANT settings.

Release note: None

#### catalog: add SystemTableIDResolver

This commit adds the SystemTableIDResolver interface and its
implementation, and modifies tenantsettingswatcher to use it for the
`tenant_settings` table.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76445)
<!-- Reviewable:end -->

RaduBerinde commented:
@ajwerner I added a commit that adds `SystemTableIDResolver` (both interface and implementation).

RaduBerinde commented:
>  I could make the host side return empty overrides if the host version is old,

I did that, I think that makes the most sense. The overrides will automatically start getting updated once the version is upgraded.

RaduBerinde committed:
logictestccl: add tenant_settings test

This commit adds functionality to logic tests to allow running SQL
against the host cluster in a tenant configuration. This allows us to
manipulate the tenant_settings table and observe the settings changing
on the tenant.

We add a tenant_settings test that verifies basic overrides behavior.
This test can become more comprehensive once we implement the new
ALTER TENANT settings.

Release note: None

RaduBerinde committed:
tenantsettingswatcher: integrate into server

This commit integrates the tenant settings watcher into the server and
implements the TenantSettings API.

Release note: None

RaduBerinde committed:
tenantsettingswatcher: implement watcher

This commit implements the tenantsettingswatcher, which monitors the
tenant_settings table and allows callers to access the current
overrides and listen for changes.

For now, we hardcode the tenant_settings system table ID to 50 (which
will be the case in fresh clusters). We will fix this in a follow-up
change.

Release note: None

RaduBerinde committed:
catalog: add SystemTableIDResolver

This commit adds the SystemTableIDResolver interface and its
implementation, and modifies tenantsettingswatcher to use it for the
`tenant_settings` table.

Release note: None

RaduBerinde committed:
tenantsettingswatcher: in-memory data structure

This commit introduces the overrides store, the in-memory data
structure that will hold all tenant overrides.

The data structure is designed for efficient read access and to allow
multiple goroutines to get notified of changes (via a channel close).

Release note: None

ajwerner commented on pkg/sql/catalog/descs/system_table.go:
```diff
@@ -0,0 +1,60 @@
+) (descpb.ID, error) {
+	var id descpb.ID
+	err := r.collectionFactory.Txn(ctx, r.ie, r.db, func(ctx context.Context, txn *kv.Txn, descriptors *Collection) error {
+		_, desc, err := descriptors.GetImmutableTableByName(
```
Now that you're here, you can use the lower-level call to get what you need. This will also avoid the lookup against kv. 

```go
var id descpb.ID
if err := r.collectionFactory.Txn(ctx, r.ie, r.db, func(
    ctx context.Context, txn *kv.Txn, descriptors *Collection,
) (err error) {
		id, err = descriptors.kv.lookupName(
    		ctx, txn, nil /* maybeDatabase */, keys.SystemDatabaseID, keys.PublicSchemaID, tableName,
		)
		return err
}); err != nil {
    return 0, err
}
return id, nil
```

ajwerner commented on pkg/server/tenantsettingswatcher/overrides_store.go:
```diff
@@ -0,0 +1,153 @@
+//
+// The caller can listen for closing of changeCh, which is guaranteed to happen
+// if the tenant's overrides change.
+func (s *overridesStore) GetTenantOverrides(tenantID roachpb.TenantID) *tenantOverrides {
```
why export this?

ajwerner commented on pkg/server/tenantsettingswatcher/overrides_store.go:
```diff
@@ -0,0 +1,153 @@
+//
+// This method is called once we complete a full initial scan of the
+// tenant_setting table.
+func (s *overridesStore) SetAll(allOverrides map[roachpb.TenantID][]roachpb.TenantSetting) {
```
why export this?

ajwerner commented on pkg/server/tenantsettingswatcher/overrides_store.go:
```diff
@@ -0,0 +1,153 @@
+	}
+}
+
+func (s *overridesStore) Init() {
```
why export this?

ajwerner commented on pkg/server/tenantsettingswatcher/overrides_store.go:
```diff
@@ -0,0 +1,153 @@
+// SetTenantOverride changes an override for the given tenant. If the setting
+// has an empty value, the existing override is removed; otherwise a new
+// override is added.
+func (s *overridesStore) SetTenantOverride(
```
why export this?

ajwerner commented on pkg/server/tenantsettingswatcher/watcher.go:
```diff
@@ -0,0 +1,286 @@
+
+// GetTenantOverrides returns the current overrides for the given tenant, and a
+// channel that will be closed when the overrides for this tenant change.
+func (w *Watcher) GetTenantOverrides(
```
Note that the returned slice is shared and must not be modified

ajwerner commented on pkg/server/tenantsettingswatcher/watcher.go:
```diff
@@ -0,0 +1,286 @@
+// GetAllTenantOverrides returns the current overrides for all tenants.
+// Each of these overrides apply to any tenant, as long as that tenant doesn't
+// have an override for the same setting.
+func (w *Watcher) GetAllTenantOverrides() (
```
Note that the returned slice is shared and must not be modified

ajwerner reviewed (approved):
LGTM

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76445)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, and @dt)

___
*[pkg/server/tenantsettingswatcher/overrides_store.go, line 63 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76445#-MveUNLeDWYY2WXN3GyA-r10-63:-MvzxFEfDkaCdcuHLc2X:bgrayg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/348cdc7732a8786316b552fe320b29f545322647/pkg/server/tenantsettingswatcher/overrides_store.go#L63)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

why export this?
</blockquote></details>

It's not really, since the type itself is not exported. It's just meant to differentiate the "API" of the datastructure from the internals. 
___
*[pkg/server/tenantsettingswatcher/watcher.go, line 271 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76445#-MveUNLf5rVGBzQ1mvfG-r10-271:-Mw-GdNr0uj8vvAm5zlh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/348cdc7732a8786316b552fe320b29f545322647/pkg/server/tenantsettingswatcher/watcher.go#L271)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Note that the returned slice is shared and must not be modified
</blockquote></details>

Done.
___
*[pkg/server/tenantsettingswatcher/watcher.go, line 281 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76445#-MveUNLf5rVGBzQ1mvfG-r10-281:-Mw-GdnX252b9konWCOL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/348cdc7732a8786316b552fe320b29f545322647/pkg/server/tenantsettingswatcher/watcher.go#L281)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Note that the returned slice is shared and must not be modified
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descs/system_table.go, line 49 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/76445#-MvjOTrVcYNbW_yqPPDL-r10-49:-Mw-H6W7DH2Pica2UQ_O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/348cdc7732a8786316b552fe320b29f545322647/pkg/sql/catalog/descs/system_table.go#L49)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Now that you're here, you can use the lower-level call to get what you need. This will also avoid the lookup against kv. 

```go
var id descpb.ID
if err := r.collectionFactory.Txn(ctx, r.ie, r.db, func(
    ctx context.Context, txn *kv.Txn, descriptors *Collection,
) (err error) {
		id, err = descriptors.kv.lookupName(
    		ctx, txn, nil /* maybeDatabase */, keys.SystemDatabaseID, keys.PublicSchemaID, tableName,
		)
		return err
}); err != nil {
    return 0, err
}
return id, nil
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76445#-:-Mw-IdHECOVl15ccs3s9:b-814g8k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @ajwerner, and @dt)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
